### DISCUSSION TITLE
dimensions of images

### DISCUSSION THREAD

**Discussion by paxo-rch** [lovyan03/LovyanGFX#301] [Q&A] [answered]
How can I get the dimensions of an image from the sd card? Lovyangfx supports bmp, jpg and png files but for me impossible to recover the dimensions? I need it to allocate a buffer of the same size. I don't know if this method already exists?

Thanks in advance

**Comment by tobozo** [comment]
Image info from the decoders (png,bmp,jpeg,qoi) is private during decoding because lgfx image functions are harmonized into a common transaction-based syntax, which is designed to be used in imperative mode.

so it's either you know the size in advance, or you speculate on it e.g. setup a sprite, draw the image in there without worrying about buffer overflows

**Reply by paxo-rch** [reply]
I already did it, it works for bmp images, but for images without background like jpg and png, it's not possible.
I can't afford to allocate a sprite of 1000x1000 just to know the dimensions of an image...

**Reply by tobozo** [reply]
understood cropping the image isn't satisfying.

I'm trying to understand the goal of copying pixels into a buffer instead of directly rendering to the display.

are you batch-creating thumbnails of images ?

**Reply by paxo-rch** [reply]
I'm creating a user interface that renders frames in ram plus displays them on screen; i need to load images in ram.

**Comment by paxo-rch** [comment, marked as answer]
```cpp
bool parsePng(String filename, int16_t* width, int16_t* height)
{
    File file = SD.open(filename);

    if (!file)
    {
        SPI_busy = false;
        return false;
    }

    byte table[32];

    for (int i = 0; file.available() && i < 32; i++)
    {
        table[i] = file.read();
    }

    *width=table[16]*256*256*256+table[17]*256*256+table[18]*256+table[19];
    *height=table[20]*256*256*256+table[21]*256*256+table[22]*256+table[23];

    file.close();

    return true;
}
```
I finally make this short code for parsing a png image and retrieve dimensions.

**Reply by tobozo** [reply]
if you don't mind the extra i/o there's an equivalent for jpeg too

```cpp
bool readJpegHeader( fs::File *file, int *width, int *height )
{
  // start of image (SOI)
  file->read();   // oxff
  file->read();   // oxd8
  // Scan miscellaneous markers until we reach SOF0 marker (0xC0)
  for(;;) {
    // next marker
    int marker;
    while((marker = file->read()) != 0xFF);
    while((marker = file->read()) == 0xFF);
    // SOF
    if (marker ==  0xC0 || marker == 0xC2) {
      file->read();   // length (2 bytes)
      file->read();   // #
      file->read();   // bpp, usually 8
      *height = (file->read() << 8) + file->read();   // height
      *width = (file->read() << 8) + file->read();   // width
      return true;
    }
  }
  return false;
}
```

**Reply by paxo-rch** [reply]
thank you very much for this time saver

**Comment by paxo-rch** [comment]
```cpp
bool parseBmp(String filename, int16_t* width, int16_t* height)
{
    File file = SD.open(filename);

    if (!file)
    {
        SPI_busy = false;
        return false;
    }

    byte table[32];

    for (int i = 0; file.available() && i < 32; i++)
    {
        table[i] = file.read();
        Serial.print(String(table[i]) + ",");
    }

    *width=table[15]*256*256*256+table[16]*256*256+table[17]*256+table[18];
    *height=table[19]*256*256*256+table[20]*256*256+table[21]*256+table[22];

    file.close();

    return true;
}```
and for the bitmap!

**Reply by tobozo** [reply]
you can boost performances by opening the file only once (doesn't need to be done by your parseImage functions **and** by drawImage )

the pre-opened file can rewinded and passed as a stream to `drawImage( Stream*)`, it's much faster than opening twice.

```cpp

int width, height;
fs::File file = fs.open("my_image_file.blah");

if( parseImage( &file, &width, &height ) ) {
   file.seek(0); // rewind
   lgfx->drawJpg( (Stream*)&file );
}

file.close();

```
